pymc_experimental/utils/cache.py

```diff
@@ -0,0 +1,188 @@
+           4. PyMC, PyTensor, and PyMC-Experimental versions
+        If a valid cache is found, sampling is bypassed altogether, unless ``force_sample=True``.
+        Otherwise, sampling is performed and the result cached for future reuse.
+        Caching is done on the basis of SHA-1 hashing, and there could be unlikely false positives.
+
+
+    Examples
+    --------
+
+    .. code-block:: python
+
+        import pymc as pm
+        from pymc_experimental.utils.cache import cache_sampling
+
+        with pm.Model() as m:
+            y_data = pm.MutableData("y_data", [0, 1, 2])
+            x = pm.Normal("x", 0, 1)
+            y = pm.Normal("y", mu=x, observed=y_data)
+
+            cache_sample = cache_sampling(pm.sample, dir="traces")
+            idata1 = cache_sample(chains=2)
+
+            # Cache hit! Returning stored result
+            idata2 = cache_sample(chains=2)
+
+            pm.set_data({"y_data": [1, 1, 1]})
+            idata3 = cache_sample(chains=2)
+
+        assert idata1.posterior["x"].mean() == idata2.posterior["x"].mean()
+        assert idata1.posterior["x"].mean() != idata3.posterior["x"].mean()
+
+    """
+    allowed_fns = (sample, sample_prior_predictive, sample_posterior_predictive)
+    if sampling_fn not in allowed_fns:
+        raise ValueError(f"Cache sampling can only be used with {allowed_fns}")
+
+    def wrapped_sampling_fn(*args, model=None, random_seed=None, **kwargs):
+        if args:
+            raise ValueError("Non-keyword arguments not allowed in cache_sampling")
+
+        extend_inferencedata = kwargs.pop("extend_inferencedata", False)
+
+        # Model hash
+        model = modelcontext(model)
+        fg, _ = fgraph_from_model(model)
+        model_hash = hash_from_fg(fg)
+
+        # Sampling hash
+        sampling_hash_dict = kwargs.copy()
+        sampling_hash_dict.pop("trace", None)
+        sampling_hash_dict.pop("random_seed", None)
+        sampling_hash_dict.pop("progressbar", None)
+        sampling_hash_dict.pop("compile_kwargs", None)
+        sampling_hash_dict["sampling_fn"] = str(sampling_fn)
+        sampling_hash_dict["versions"] = (
+            pymc.__version__,
+            pytensor.__version__,
+            pymc_experimental.__version__,
+        )
+        sampling_hash = str(sampling_hash_dict)
+
+        file_name = hashlib.sha1((model_hash + sampling_hash).encode()).hexdigest() + ".nc"
+        file_path = os.path.join(dir, file_name)
+
+        if not force_sample and os.path.exists(file_path):
+            print("Cache hit! Returning stored result", file=sys.stdout)
+            idata_out: az.InferenceData = az.from_netcdf(file_path)
+            if force_load:
+                idata_out.load()
+
+        else:
+            idata_out = sampling_fn(*args, **kwargs, model=model, random_seed=random_seed)
+            if os.path.exists(file_path):
+                os.remove(file_path)
+            if not os.path.exists(dir):
+                os.mkdir(dir)
```

Comment:
```suggestion
                os.makedirs(dir, exist_ok=True)
```

I think that it's better to use [`os.makedirs`](https://docs.python.org/3/library/os.html#os.makedirs) because it creates the intermediate directories if they are required.